### ISSUE TITLE
AuthManager's getUserToken only tries to obtain authentication code, not refresh token

### ISSUE THREAD

**Issue by Sellular** [spotify-api/spotify-api.js#32]
# Bug Report

## My version of spotify-api.js
8.0.0

## What i expected...
Using the getUserToken() method in AuthManager with refresh_token passed, I would expect to get a new access and refresh token

## But what happened is...
What happened instead is I get an error that says 'Invalid authentication code', making me believe that it is requesting an authentication code instead of a refresh token. Upon further inspection into the source code this seems to be the case, as it just populates the `code` field. Maybe this is what the spotify API wanted before, but has since changed? I'm not sure

Sadly, I don't have the code I was using anymore since I switched to a different spotify API node layer

**Comment by Scientific-Guy** [comment]
I can understand what you mean but still its good if i get the code which has thrown the error can help me to find the bug.

**Comment by Sellular** [comment]
> I can understand what you mean but still its good if i get the code which has thrown the error can help me to find the bug.

Ill see if I can pull up a dummy project and recreate it at some point, may be a day or two

**Comment by Scientific-Guy** [comment]
And 4 days has been passed without a error producing code i cannot clearly say what you are actually meaning.

**Comment by Sellular** [comment]
Here's some sample code. I login by hitting the /login endpoint, take the refresh_token that is sent back in the /callback response after logging in and use that in the next run of the script. It gives the error "Invalid authorization code" upon trying to login. Sorry it took a few days, I was on vacation.

`const Spotify = require('spotify-api.js');
const express = require('express');
require('dotenv').config();

const app = express();
const spotifyClient = new Spotify.Client();

spotifyClient.login({
    clientID: process.env.CLIENT_ID,
    clientSecret: process.env.CLIENT_SECRET,
    redirectURL: process.env.REDIRECT_URI,
    refreshToken: process.env.REFRESH_TOKEN 
});

// Login site
app.get('/login', (req, res) => {
    res.redirect("https://accounts.spotify.com/en/authorize?" +
        "client_id=" +
        process.env.CLIENT_ID +
        "&" +
        "redirect_uri=" +
        encodeURIComponent(process.env.REDIRECT_URI) +
        "&" +
        "response_type=code" +
        "&" +
        "scope=" +
        encodeURIComponent('user-top-read')
    );
});

app.get('/callback', async (req, res) => {

    const auth = await spotifyClient.login({
        clientID: process.env.CLIENT_ID, // Your spotify app client id
        clientSecret: process.env.CLIENT_SECRET, // Your spotify app client secret
        redirectURL: process.env.REDIRECT_URI, // Will not redirect, it is a verification method so keep the same redirect uri
        code: req.query.code // Code sent by the api
    }); // Short hand method

    res.send({
        auth
    }); // Returns the user information!
    // If you have the required scopes you can also access the user's player, etc!

});

app.listen(3000);`

**Comment by Sellular** [comment]
[](url)

**Comment by Sellular** [comment]
Didn't mean to close, reopening

**Comment by Scientific-Guy** [comment]
> Here's some sample code. I login by hitting the /login endpoint, take the refresh_token that is sent back in the /callback response after logging in and use that in the next run of the script. It gives the error "Invalid authorization code" upon trying to login. Sorry it took a few days, I was on vacation.
> 
> `const Spotify = require('spotify-api.js');
> const express = require('express');
> require('dotenv').config();
> 
> const app = express();
> const spotifyClient = new Spotify.Client();
> 
> spotifyClient.login({
> clientID: process.env.CLIENT_ID,
> clientSecret: process.env.CLIENT_SECRET,
> redirectURL: process.env.REDIRECT_URI,
> refreshToken: process.env.REFRESH_TOKEN
> });
> 
> // Login site
> app.get('/login', (req, res) => {
> res.redirect("https://accounts.spotify.com/en/authorize?" +
> "client_id=" +
> process.env.CLIENT_ID +
> "&" +
> "redirect_uri=" +
> encodeURIComponent(process.env.REDIRECT_URI) +
> "&" +
> "response_type=code" +
> "&" +
> "scope=" +
> encodeURIComponent('user-top-read')
> );
> });
> 
> app.get('/callback', async (req, res) => {
> 
> ```
> const auth = await spotifyClient.login({
>     clientID: process.env.CLIENT_ID, // Your spotify app client id
>     clientSecret: process.env.CLIENT_SECRET, // Your spotify app client secret
>     redirectURL: process.env.REDIRECT_URI, // Will not redirect, it is a verification method so keep the same redirect uri
>     code: req.query.code // Code sent by the api
> }); // Short hand method
> 
> res.send({
>     auth
> }); // Returns the user information!
> // If you have the required scopes you can also access the user's player, etc!
> ```
> 
> });
> 
> app.listen(3000);`

The code block looks broken but the code itself looks broken. But if i am right the dummy code looks something like this

```js
const Spotify = require('spotify-api.js');
const express = require('express');
require('dotenv').config();

const app = express();
const spotifyClient = new Spotify.Client();

spotifyClient.login({
     clientID: process.env.CLIENT_ID,
     clientSecret: process.env.CLIENT_SECRET,
     redirectURL: process.env.REDIRECT_URI,
     refreshToken: process.env.REFRESH_TOKEN
});

// Login site
app.get('/login', (req, res) => {
     res.redirect("https://accounts.spotify.com/en/authorize?" +
         "client_id=" +
         process.env.CLIENT_ID +
        "&redirect_uri=" +
         encodeURIComponent(process.env.REDIRECT_URI) +
        "&response_type=code&scope=" +
         encodeURIComponent('user-top-read')
     );
});

app.get('/callback', async (req, res) => {
    const auth = await spotifyClient.login({
        clientID: process.env.CLIENT_ID, // Your spotify app client id
        clientSecret: process.env.CLIENT_SECRET, // Your spotify app client secret
        redirectURL: process.env.REDIRECT_URI, // Will not redirect, it is a verification method so keep the same redirect uri
        code: req.query.code // Code sent by the api
    }); // Short hand method

    res.send({ auth });
});

app.listen(3000);`
```

First does the env has all the values right? And remember this part is an async operation

```js
spotifyClient.login({
     clientID: process.env.CLIENT_ID,
     clientSecret: process.env.CLIENT_SECRET,
     redirectURL: process.env.REDIRECT_URI,
     refreshToken: process.env.REFRESH_TOKEN
});
```

You should catch it something like

```js
spotifyClient.login({
     clientID: process.env.CLIENT_ID,
     clientSecret: process.env.CLIENT_SECRET,
     redirectURL: process.env.REDIRECT_URI,
     refreshToken: process.env.REFRESH_TOKEN
})
.then(response => {
    if (typeof response == "object") {
        // New refresh token
        process.env.REFRESH_TOKEN = response.refreshToken;
        console.log("Found a  refresh token.");
   } else console.log("Not refresh token found.")
})
.catch(console.log);
```

Just check.

**Comment by Sellular** [comment]
Not sure why the code formatting is weird.

I'm running the script two different times and using dotenv to hold the
process.env variables. Yes I'm positive the values are right. I'm not using
any .then()s cause they aren't necessary in this case, although I could.
The first time the script runs I don't have the first .login() in the code,
just the express app, sorry I didn't clarify that initially. That callback
function returns the refresh_token, which I then put into the .env file,
uncomment the first .login() that uses the refresh token, re-run the script
so it tries to login and I get the error.

**Comment by Scientific-Guy** [comment]
No here awaiting the promise is necessary because in promise you might be attempting to login before the promise gets completed and saves the token. Sometimes this would work as it works but sometimes the promise might get rejected and you would not even know that thing. So check once if it actually passes or gets rejected and kindly know that the refreshToken can be used only once.

**Comment by Sellular** [comment]
> No here awaiting the promise is necessary because in promise you might be attempting to login before the promise gets completed and saves the token. Sometimes this might work without awaiting it still sometimes the promise can get rejected and you would not even know that thing. So check once if it actually passes or gets rejected and kindly know that the refreshToken can be used only once.
> 
> And you try if the token is ever loaded for example 
> 
> ```js
> console.log(client.token);
> console.log(client.token == "NO TOKEN"); // client.token is set to "NO TOKEN" by default. If it shows true then it means the token is not loaded yet.
> ``

There's no need to await the login that I'm passing the refresh token to as it's ran on the script startup and doesn't have any dependants after. I know how async and await/then works, and this is definitely an issue with this project's code. I have the first login occurrence (one using the refresh token) commented out during the first run, since you won't have a refresh token. I hit the login, login, and get the AuthRefresh in the callback once login is complete. Then, you can either leave the login/callback statements, or comment them out, and uncomment the login w/ refresh token. It will try and login using the refresh token and it throws the error as it's trying to use an auth code and not the refresh token in the underlying code of this project.

**Comment by Scientific-Guy** [comment]
> > No here awaiting the promise is necessary because in promise you might be attempting to login before the promise gets completed and saves the token. Sometimes this might work without awaiting it still sometimes the promise can get rejected and you would not even know that thing. So check once if it actually passes or gets rejected and kindly know that the refreshToken can be used only once.
> > And you try if the token is ever loaded for example
> > ```js
> > console.log(client.token);
> > console.log(client.token == "NO TOKEN"); // client.token is set to "NO TOKEN" by default. If it shows true then it means the token is not loaded yet.
> > ``
> > ```
> 
> There's no need to await the login that I'm passing the refresh token to as it's ran on the script startup and doesn't have any dependants after. I know how async and await/then works, and this is definitely an issue with this project's code. I have the first login occurrence (one using the refresh token) commented out during the first run, since you won't have a refresh token. I hit the login, login, and get the AuthRefresh in the callback once login is complete. Then, you can either leave the login/callback statements, or comment them out, and uncomment the login w/ refresh token. Rerunning the newly commented/uncommented script will try and login using the refresh token and it throws the error as it's trying to use an auth code and not the refresh token in the underlying code of this project.

But did `client.token` loaded at the first place? So i could find where's the problem.

**Comment by Sellular** [comment]
I'll check when I'm off work

**Comment by Sellular** [comment]
@Scientific-Guy So it doesn't seem that client.token isn't being loaded, but if I am supplying the refresh token in the login method, the client.token shouldn't be populated anyway, right? If I am asking the client to login, supplying the refresh_token, it should be supplying and populating the client.token AFTER the login. Correct? Sorry for the late reply, it's been a busy week

**Comment by Scientific-Guy** [comment]
Yes client.token shall have value after login. So the problem is actually with the login code. Because you are not the only guy who faced this error. It will be good if you give more information about the bug. This would be probably fixed it in the bug fix v8.1.

**Comment by Sellular** [comment]
> Yes client.token shall have value after login. So the problem is actually with the login code. Because you are not the only guy who faced this error. It will be good if you give more information about the bug. This would be probably fixed it in the bug fix v8.1.

Just let me know if you need any more info

**Comment by Scientific-Guy** [comment]
Yeah so sorry for the very late response but i solved the bug. You can use the latest release [v8.1.0](https://github.com/spotify-api/spotify-api.js/releases/tag/8.1.0). If you still find some bugs reopen this issue or create a new one.
